python_modules/dagster/dagster/_core/definitions/asset_graph_differ.py

```diff
@@ -116,55 +182,122 @@ def from_external_repositories(
             # for dependencies have changed since ChangeReason.DEPENDENCIES is already in the list of changes
             for upstream_asset in branch_asset.parent_keys:
                 if self.branch_asset_graph.get_partition_mapping(
                     asset_key, upstream_asset
                 ) != self.base_asset_graph.get_partition_mapping(asset_key, upstream_asset):
-                    changes.append(AssetDefinitionChangeType.DEPENDENCIES)
+                    if include_details:
+                        changes.append(
+                            AssetDefinitionChangeDetailDependencies(changed={upstream_asset})
+                        )
+                    else:
+                        changes.append(AssetDefinitionChangeDetailDependencies())
                     break
 
         if branch_asset.partitions_def != base_asset.partitions_def:
-            changes.append(AssetDefinitionChangeType.PARTITIONS_DEFINITION)
+            if include_details:
+                changes.append(
+                    AssetDefinitionChangeDetailPartitionsDefinition(
+                        base_value=type(base_asset.partitions_def).__name__
+                        if base_asset.partitions_def
+                        else None,
+                        branch_value=type(branch_asset.partitions_def).__name__
+                        if branch_asset.partitions_def
+                        else None,
+                    )
+                )
+            else:
+                changes.append(AssetDefinitionChangeDetailPartitionsDefinition())
 
         if branch_asset.tags != base_asset.tags:
-            changes.append(AssetDefinitionChangeType.TAGS)
+            if include_details:
+                added, changed, removed = self._get_map_keys_diff(
+                    base_asset.tags, branch_asset.tags
+                )
+                changes.append(
+                    AssetDefinitionChangeDetailTags(
+                        added=list(added), changed=list(changed), removed=list(removed)
+                    )
+                )
+            else:
+                changes.append(AssetDefinitionChangeDetailTags())
 
         if branch_asset.metadata != base_asset.metadata:
-            changes.append(AssetDefinitionChangeType.METADATA)
+            if include_details:
+                added, changed, removed = self._get_map_keys_diff(
+                    base_asset.metadata, branch_asset.metadata
+                )
+                changes.append(
+                    AssetDefinitionChangeDetailMetadata(
+                        added=list(added), changed=list(changed), removed=list(removed)
+                    )
+                )
+            else:
+                changes.append(AssetDefinitionChangeDetailMetadata())
 
         return changes
 
+    def _get_map_keys_diff(
+        self, base: Mapping[str, Any], branch: Mapping[str, Any]
+    ) -> Tuple[Set[str], Set[str], Set[str]]:
```

Comment:
nit: in cases where we return multiple values, especially when they're the same type like they are here, it can be nice to define a `NamedTuple` and return that instead - could easily see accidentally mixing up the order of the removed/changed return values for example

```python
class MappedKeyDiff(NamedTuple):
  added: Set[str]
  changed: Set[str]
  removed: Set[str]

def _get_map_keys_diff(...) -> MappedKeyDiff:
  ...
  return MappedKeyDiff(added, changed, removed)
```